packages/next-swc/crates/next-api/src/dynamic_imports.rs

```diff
@@ -0,0 +1,395 @@
+///    - Server embeds those into __NEXT_DATA__ and sent to the client. (https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/server/render.tsx#L1453)
+///    - When client boots up, pass it to the client preload (https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/client/index.tsx#L943)
+///    - Loadable runtime injects preload fn to wait until all the dynamic components are being loaded, this ensures hydration mismatch won't occur
+///      (https://github.com/vercel/next.js/blob/ad42b610c25b72561ad367b82b1c7383fd2a5dd2/packages/next/src/shared/lib/loadable.shared-runtime.tsx#L281)
+pub(crate) async fn collect_next_dynamic_imports(
+    entry: Vc<Box<dyn EcmascriptChunkPlaceable>>,
+) -> Result<IndexMap<Vc<Box<dyn Module>>, DynamicImportedModules>> {
+    // Traverse referenced modules graph, collect all of the dynamic imports:
+    // - Read the Program AST of the Module, this is the origin (A)
+    //  - If there's `dynamic(import(B))`, then B is the module that is being
+    //    imported
+    // Returned import mappings are in the form of
+    // (Module<A>, Vec<(B, Module<B>)>) (where B is the raw import source string,
+    // and Module<B> is the actual resolved Module)
+    let imported_modules_mapping = NonDeterministic::new()
+        .skip_duplicates()
+        .visit([Vc::upcast(entry)], get_referenced_modules)
+        .await
+        .completed()?
+        .into_inner()
+        .into_iter()
+        .map(build_dynamic_imports_map_for_module);
+
+    // Consolidate import mappings into a single indexmap
+    let mut import_mappings: IndexMap<Vc<Box<dyn Module>>, DynamicImportedModules> =
+        IndexMap::new();
+
+    for module_mapping in imported_modules_mapping {
+        if let Some(module_mapping) = &*module_mapping.await? {
+            let (origin_module, dynamic_imports) = &*module_mapping.await?;
+            import_mappings
+                .entry(*origin_module)
+                .or_insert_with(Vec::new)
+                .append(&mut dynamic_imports.clone())
+        }
+    }
+
+    Ok(import_mappings)
+}
+
+async fn get_referenced_modules(
+    parent: Vc<Box<dyn Module>>,
+) -> Result<impl Iterator<Item = Vc<Box<dyn Module>>> + Send> {
+    primary_referenced_modules(parent)
+        .await
+        .map(|modules| modules.clone_value().into_iter())
+}
+
+#[turbo_tasks::function]
+async fn build_dynamic_imports_map_for_module(
+    module: Vc<Box<dyn Module>>,
+) -> Result<Vc<OptionDynamicImportsMap>> {
+    let Some(ecmascript_asset) =
+        Vc::try_resolve_downcast_type::<EcmascriptModuleAsset>(module).await?
+    else {
+        return Ok(OptionDynamicImportsMap::none());
+    };
+
+    let ParseResult::Ok { program, .. } = &*ecmascript_asset.parse().await? else {
+        NextDynamicParsingIssue {
+            ident: module.ident(),
+        }
+        .cell()
+        .emit();
```

Comment:
This error will have already been emitted (not specific to loadable, but the parse error would have) during the graph construction. I think we can just bail here, no need to emit another error.

Reply:
if we bail we fail entire rendering, in result takes down next-dev and makes test case execution fails. I originally bailed out then fall back to this.

Reply:
Sorry, not literally `bail!`, just return a `none()` without emitting.